File: src/main/java/dragonbook/inter/stmt/Else.java
package dragonbook.inter.stmt;

import dragonbook.inter.expr.Expr;
import dragonbook.symbols.Type;

public class Else extends Stmt {
    public final Expr condition;
    public final Stmt then;
    public final Stmt elseStmt;

    public Else(Expr condition, Stmt then, Stmt elseStmt) {
        if (condition.type != Type.BOOL) condition.error("boolean required in if");
        this.condition = condition;
        this.then = then;
        this.elseStmt = elseStmt;
    }
}

File: src/main/kotlin/dragonbook/inter/stmt/Else.kt
package dragonbook.inter.stmt

import dragonbook.error.ParseError
import dragonbook.inter.expr.Expr
import dragonbook.symbols.Type

data class Else(val condition: Expr, val then: Stmt, val elseStmt: Stmt) : Stmt {

    init {
        validate()
    }

    private fun validate() {
        if (condition.type != Type.BOOL) throw ParseError("boolean required in if", condition)
    }

    override fun toString(): String = "if ($condition) { $then } else { $elseStmt }"
}